### PR TITLE
Underlying plumbing/infrastructure for Porter CLI/RestApp

### LINKED ISSUE #2668: Basic Porter "Learner"
Porter needs to learn about the nucypher network and ursulas in order to execute network functionality on behalf of Alice(#2666) and Bob (#2667).

Presumably, this makes Porter some sort of a `Learner`:
* Some analogies may be drawn from the status monitor "Crawler" - https://github.com/nucypher/nucypher-monitor/blob/main/monitor/crawler.py#L152
* Stubbed client functions to be used by eventual endpoints (functions to be actually implemented with associated endpoints in other issues) - this just provides the basic plumbing

### LINKED ISSUE #2702: REST App/CLI built around Porter "Learner"
Related to #2668 
* REST APP plumbing for Porter Learner
* Stubbed REST endpoints for Alice and Bob - only stubbed not implemented i.e. only plumbing - endpoints and associated Porter Learner function implementation can be completed in separate issues
* Porter CLI to instantiate and run web app

### PR DESCRIPTION
**Pull request by derekpierre**
**Type of PR:**
- [ ] Bugfix
- [ ] Feature
- [ ] Documentation
- [X] Other 

**Required reviews:** 
- [ ] 1
- [ ] 2
- [X] 3

**What this does:**
This refactoring will facilitate a non-character learner like "Porter" to utilize the existing schema/controllers infrastructure for REST-like endpoints. 

A lot of prior work was done for CLI/JSON/Web controllers and verification of inputs via schema definitions for the "character control" that can be reused for Porter. However, Porter isn't a character, so some of the plumbing needs to be modified to allow for the possibility of a non-character utilizing the same infrastructure. I think of it as just a base control for that maps to an interface and the underlying implementation of that interface may be a character (eg. Alice, Bob etc.) or not (eg. Porter).

Chatted with @vepkenez and @KPrasch about it yesterday.

Once that refactoring is done, Porter can benefit from the ease of utilizing a CLIConroller and a WebController with built-in function mapping and parameter validation.

Closes #2702 .
Closes #2668.

### CODE REVIEW

**Comment by derekpierre** [comment]
There is now a simple CLI available and a WebController. Of course, any actual Porter functionality on behalf of Alice/Bob is only stubbed out and not implemented. However, the control infrastructure allows for implicit CLI options, and rest endpoints to be appropriately mapped.
 
```bash
$ nucypher porter --help
Usage: nucypher porter [OPTIONS] COMMAND [ARGS]...

  Porter management commands. Porter is the conduit between web apps and the
  nucypher network, that performs actions on behalf of Alice and Bob.

Options:
  --help  Show this message and exit.

Commands:
  exec-work-order       Execute a PRE work order on behalf of Bob.
  get-treasure-map      Retrieve a treasure map on behalf of Bob.
  get-ursulas           Sample Ursulas on behalf of Alice.
  publish-treasure-map  Publish a treasure map on behalf of Alice.
  revoke                Off-chain revoke of a policy on behalf of Alice.
  run                   Start Porter's Web controller.
```

And a very simple HTTP Web controller that can be run:
```
$ nucypher porter run --provider https://mainnet.infura.io/v3/<app_id> --network mainnet


 ______
(_____ \           _
 _____) )__   ____| |_  ____  ____
|  ____/ _ \ / ___)  _)/ _  )/ ___)
| |   | |_| | |   | |_( (/ /| |
|_|    \___/|_|    \___)____)_|

the conduit between web apps and the nucypher network

Network: Mainnet
Provider: https://mainnet.infura.io/v3/<app_id>
Running Porter Web Controller at http://localhost:9155
```

(In parallel)
```
$ curl http://localhost:9155/get_ursulas
{'quantity': ['Missing data for required field.'], 'duration_periods': ['Missing data for required field.']}
```

File: nucypher/control/controllers.py [hunk 1]
```diff
@@ -0,0 +1,341 @@
...
+    """
+    A transactional interface for a human to interact with.
+    """
+    _emitter_class = NotImplemented
+
+    def __init__(self):
+        # Control Emitter
+        self.emitter = self._emitter_class()
+
+    def _perform_action(self, action: str, request: dict) -> dict:
```
Comment (piotr-roslaniec) [inline nucypher/control/controllers.py:49]: `Optional[dict] = None`?

File: nucypher/control/controllers.py [hunk 2]
```diff
@@ -0,0 +1,341 @@
...
+        """
+        This method is where input validation and method invocation
+        happens for all interface actions.
+        """
+        request = request or {}  # for requests with no input params request can be ''
```
Comment (piotr-roslaniec) [inline nucypher/control/controllers.py:54]: Should we add some sort of validation here to make sure that certain `action`s have the required `request`s?

Comment (derekpierre) [reply nucypher/control/controllers.py:54]: Great question. This is the beauty of the original "control" concept that I've used as a base. There is implicit validation that is performed on requests via a combination of the `marshmallow.fields` and `BaseSchema` concepts

The use of the `BaseSchema` concept defines the fields that make up the request, and the `marshmallow.fields` defines the validation for each field type. 

For the Schema definitions see:
- nucypher.utilities.porter.control.specifications.porter_schema.py (Porter)
- nucypher.characters.control.specifications.(alice.py, bob.py, enrico.py); for the original individual character schemas

For field validation see:
- nucypher.control.specifications.fields (for basic fields)
- nucypher.characters.control.specifications.fields (for fields needed for the character requests)
- nucypher.utilities.porter.control.specifications.fields (for fields needed  for Porter requests)

File: nucypher/control/controllers.py (hunk 1 above)
Comment (derekpierre) [reply nucypher/control/controllers.py:49]: ✔️

**Comment by derekpierre** [comment]
Filed follow-up issue #2714 for reviewing the inconsistencies with some of the original character controls fields.

File: nucypher/blockchain/eth/actors.py [hunk 3]
```diff
@@ -100,7 +100,7 @@
     prettify_eth_amount
 )
 from nucypher.characters.banners import STAKEHOLDER_BANNER
-from nucypher.characters.control.emitters import StdoutEmitter
+from nucypher.control.emitters import StdoutEmitter
```
Comment (KPrasch) [inline nucypher/blockchain/eth/actors.py:103]: 👍🏻

File: nucypher/characters/base.py [hunk 4]
```diff
@@ -524,9 +525,9 @@ def make_rpc_controller(self, crash_on_error: bool = False):
 
     def make_cli_controller(self, crash_on_error: bool = False):
         app_name = bytes(self.stamp).hex()[:6]
-        controller = CLIController(app_name=app_name,
-                                   crash_on_error=crash_on_error,
-                                   interface=self.interface)
+        controller = CharacterCLIController(app_name=app_name,
```
Comment (KPrasch) [inline nucypher/characters/base.py:528]: Good preservation of the namespacing.

File: nucypher/cli/main.py [hunk 5]
```diff
@@ -86,6 +87,7 @@ def nucypher_cli():
     felix.felix,        # Faucet
     cloudworkers.cloudworkers,  # Remote Worker node management
     contacts.contacts,          # Character "card" management
+    porter.porter
```
Comment (KPrasch) [inline nucypher/cli/main.py:90]: 💥

File: nucypher/control/specifications/fields/base.py [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+    def _validate(self, value):
+        if not value > 0:
+            raise InvalidInputData(f"{self.name} must be a positive integer.")
+
+
+#
+# CLI utility
+#
+
+class click:
```
Comment (KPrasch) [inline nucypher/control/specifications/fields/base.py:59]: What's going on here?

File: nucypher/utilities/porter/control/controllers.py [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+                 *args,
+                 **kwargs):
+        super().__init__(interface=interface, *args, **kwargs)
+
+    def _perform_action(self, *args, **kwargs) -> dict:
+        try:
+            response_data = super()._perform_action(*args, **kwargs)
+        finally:
+            self.log.debug(f"Finished action '{kwargs['action']}', stopping {self.interface.implementer}")
+            self.interface.implementer.disenchant()
```
Comment (KPrasch) [inline nucypher/utilities/porter/control/controllers.py:37]: 🌷

File: nucypher/utilities/porter/porter.py [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+    BANNER = r"""
+
+ ______
+(_____ \           _
+ _____) )__   ____| |_  ____  ____
+|  ____/ _ \ / ___)  _)/ _  )/ ___)
+| |   | |_| | |   | |_( (/ /| |
+|_|    \___/|_|    \___)____)_|
+
+the conduit between web apps and the nucypher network
```
Comment (KPrasch) [inline nucypher/utilities/porter/porter.py:39]: Not very catchy 😆

File: nucypher/utilities/porter/porter.py [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+
+from nucypher.blockchain.eth.registry import BaseContractRegistry, InMemoryContractRegistry
+from nucypher.control.controllers import WebController, JSONRPCController
+from nucypher.network.nodes import Learner
+from nucypher.utilities.logging import Logger
+from nucypher.utilities.porter.control.controllers import PorterCLIController
+from nucypher.utilities.porter.control.interfaces import PorterInterface
+
+
+class Porter(Learner):
```
Comment (KPrasch) [inline nucypher/utilities/porter/porter.py:28]: Naming things: Are we settled on "Porter"?

File: nucypher/utilities/porter/porter.py [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+
+        super().__init__(save_metadata=True, domain=domain, *args, **kwargs)
+
+        self.log = Logger(self.__class__.__name__)
+
+        # Controller Interface
+        self.interface = self._interface_class(porter=self)
+        self.controller = NO_CONTROL_PROTOCOL
+        if controller:
+            # TODO need to understand this better - only made it analogous to what was done for characters
```
Comment (KPrasch) [inline nucypher/utilities/porter/porter.py:68]: Is there any scernerio where you'd want to programmatically use a Porter instance *without* the CLI?

File: nucypher/utilities/porter/porter.py [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+        porter_flask_control = controller.make_control_transport()
+
+        #
+        # Porter Control HTTP Endpoints
+        #
+        @porter_flask_control.route('/get_ursulas', methods=['GET'])
+        def get_ursulas() -> Response:
+            """
+            Porter control endpoint for sampling Ursulas on behalf of Alice.
+            """
```
Comment (KPrasch) [inline nucypher/utilities/porter/porter.py:103]: Style: These docstrings can all be one-liners.

File: nucypher/control/specifications/fields/base.py (hunk 6 above)
Comment (derekpierre) [reply nucypher/control/specifications/fields/base.py:59]: I believe it is part of the mechanism by which the CLI commands auto-derive cli options from Schemas, see https://github.com/derekpierre/nucypher/blob/control-refactoring/nucypher/control/interfaces.py#L47 and https://github.com/derekpierre/nucypher/blob/control-refactoring/nucypher/control/specifications/fields/base.py#L29. That way the options are tied to schemas and don't need to be redefined in the CLI.

Basically, I continued the use from the previous character control - perhaps @vepkenez can clarify better than I can.

File: nucypher/utilities/porter/porter.py (hunk 11 above)
Comment (derekpierre) [reply nucypher/utilities/porter/porter.py]: ✔️

File: nucypher/control/controllers.py (hunk 2 above)
Comment (piotr-roslaniec) [reply nucypher/control/controllers.py:58]: Thanks for the explanation, that's a really cool take on schema validation :+1:

File: nucypher/control/specifications/fields/base.py (hunk 6 above)
Comment (vepkenez) [reply nucypher/control/specifications/fields/base.py:59]: this line and class make no sense... `click` is already defined as it is imported already (and this was already happening before you cleaned this up @derekpierre.
I think we can probably get rid of it.  It must have been something I started but then reworked.

File: nucypher/control/controllers.py (hunk 2 above)
Comment (derekpierre) [reply nucypher/control/controllers.py:58]: It's pretty cool. The original "control" concept was created by @vepkenez and @KPrasch - I'm just refactoring it to make it non-character specific.

File: nucypher/control/specifications/fields/base.py (hunk 6 above)
Comment (derekpierre) [reply nucypher/control/specifications/fields/base.py:59]: ✔️

File: nucypher/utilities/porter/porter.py (hunk 8 above)
Comment (derekpierre) [reply nucypher/utilities/porter/porter.py]: ✔️ - temporarily updated. Can be changed later on if needed.

**Review by KPrasch — APPROVED** [review]

**Review by piotr-roslaniec — APPROVED** [review]

**Review by vepkenez — APPROVED** [review]
